### ISSUE TITLE
embedding of "resps" (cont.)

### ISSUE THREAD

**Issue by frank613** [e-c-k-e-r/vall-e#7]
Hi Ecker,

Thank you for the clarification yesterday. But I felt I have trouble in understanding that part again. 

```
		embedding = self.resps_emb(
								input if input.dim() == 1 or quant_level == 0 else input[:, :quant_level],
								#offset = 0 if classifier_level.startswith("AR:") else 1,
								name = classifier_level,
								quant_level = 0 if quant_level == 0 else quant_level - 1, 
							)

```
Let's say we use the Encodec which has resps_level [1-8] and the corresponding quant_level range is [0-7]. If we are modelling quant_level = 2, then we will need the sum of embeddings from NAR:0:1 and NAR:1:2.  However, the code above passes  quant_level - 1 = 1 to the "Audioembedding", and according to the function forward() of "Audioembeding", it will only iterate once in the list comprehension which takes only from NAR:0:1 as embeddings for modeling/predicting code of quant_level 2:

```
if self.sums and quant_level > 0:
			x = sum( [ self.embeddings[k + offset]( xi[:, k] ) for k in range( quant_level ) ] )
```

Did I miss anything from your code?

Regards,
Xinwei

**Comment by e-c-k-e-r** [comment]
> Did I miss anything from your code?

I imagine it's from the confusing nature of both having to offset to line things up, plus the mismatching definition of `quant_level` inside and outside of `AudioEmbedding`.

For `ar+nar-len-llama-8`'s implementation, the `AudioEmbedding.name`s map to the corresponding `AudioEmbedding.embeddings[]` weight (where its position in the array maps to the position in the embeddings list):
```
{
  "AR:0:0": "resps_emb.embeddings[0]",
  "NAR:0:1": "resps_emb.embeddings[1]",
  "NAR:1:2": "resps_emb.embeddings[2]",
  "NAR:2:3": "resps_emb.embeddings[3]",
  "NAR:3:4": "resps_emb.embeddings[4]",
  "NAR:4:5": "resps_emb.embeddings[5]",
  "NAR:5:6": "resps_emb.embeddings[6]",
  "NAR:6:7": "resps_emb.embeddings[7]",
  "NAR:0:0": "resps_emb.embeddings[8]",
}
```

Invoking `resps_emb.forward(input, name="AR:0:0", quant_level=0)` will embed `input` through `resps_emb.embeddings[0]`, as there's no summing required.

Invoking `resps_emb.forward(input, name="NAR:1:2", quant_level=2)` will embed `input` through:
```
sum([
  resps_emb.embeddings[1]( input[:, 0] ),
  resps_emb.embeddings[2]( input[:, 1] ),
])
```
as `offset` will be set to the starting embedding index at `NAR:1:2`, minus the inputted codebook levels.

Invoking `resps_emb.forward(input, name="NAR:6:7", quant_level=7)` will embed `input` through:
```
sum([
  resps_emb.embeddings[1]( input[:, 0] ),
  resps_emb.embeddings[2]( input[:, 1] ),
  resps_emb.embeddings[3]( input[:, 2] ),
  resps_emb.embeddings[4]( input[:, 3] ),
  resps_emb.embeddings[5]( input[:, 4] ),
  resps_emb.embeddings[6]( input[:, 5] ),
  resps_emb.embeddings[7]( input[:, 6] ),
])
```

The `quant_level = [...] quant_level - 1` is because the assigned target RVQ codebook level is one higher than the `quant_level` the `AudioEmbeddings` class will see through the input, as we're predicting the target codebook level, but only have `quant_level - 1` codebooks to inference with. Inferencing works out because the `AR_NAR.forward_ar`/`AR_NAR.forward_nar` functions can both explicitly pass in the target `quant_level` at inference time, or imply it through `resps_list[#].shape[-1]`.

As I mentioned it's quite the mess of code, but I can push a commit to better clarify things in the variable names.

**Comment by frank613** [comment]
> Invoking resps_emb.forward(input, name="NAR:1:2", quant_level=2) will embed input through:
> ....

If resps_emb.forward(input, name="NAR:1:2", quant_level=2) is invoked, it means that outside the  "AudioEmbedding" we need classifier_level = "NAR:1:2" and quant_level=3, but how come we have those two pairs in the "inputs" as the code shown below will have to generate classifier_level = "NAR:2:3" according to quant_level 3.

```
if self.rvq_l_emb is not None and not self.interleave:
	inputs[i].append( ( "quant_level", torch.tensor([ quant_level ], device=device, dtype=torch.int16) ) )

	classifier_level = "AR:0:0" if quant_level == 0 else f'NAR:{quant_level-1}:{quant_level}'
```

I think I am still confused with the quant_levels. If we are predicting the third RVQ level out of eighth (3/8), then which quant_level should we use 2 or 3? What is the range of quant_level in case of 8 code books? Is it [1,8] or [0,7] or [0,8]

Thanks,
Xinwei

**Comment by frank613** [comment]
> Invoking resps_emb.forward(input, name="NAR:1:2", quant_level=2) will embed input through:
> 
> sum([
>   resps_emb.embeddings[1]( input[:, 0] ),
>   resps_emb.embeddings[2]( input[:, 1] ),
> ])

If i understand correctly, because the index of "NAR:1:2" is 2, then the offset becomes 0 as (index-quant_level=0), the return value would instead become sum( resps_emb.embeddings**[0]**( input[:, 0] ), resps_emb.embeddings**[1]**( input[:, 1] ))

Thanks,
Xinwei

**Comment by e-c-k-e-r** [comment]
> If i understand correctly, because the index of "NAR:1:2" is 2, then the offset becomes 0 as (index-quant_level=0)

Oops; in that proposed example, `resps_emb.forward(input, name="NAR:1:2", quant_level=2)` should be `quant_level=1`, as the input only has 2 codebooks to embed. Having to mentally do these 0-index / swapping between input/target codebook level confused me there.
If `quant_level=2` *were* to be passed in that example, then it would use the wrong offsets from shifting back by `quant_level`.

It might be better to instead just rely on implicitly deriving the input `quant_level` from the input tensor shape, but that could break a few things, such as the `proms_emb` that could take in all eight codebooks but is targetting a lower `quant_level` for training/inferencing.

> But I am not sure if "NAR:2:3" is semantically correct as you explained earlier because we are here handling the embedding RVQ code book 1 to predict RVQ code book 2

The confusion is from referring to nth codebook level when in-code they're 0th indexed, where `quant_level=1` refers to the second codebook in the audio, `quant_level=2` refers to the third codebook in the audio, etc.

In an ideal cleanup/rewrite I should have this mess of terminology cleared up, things like:
* `quant_level` vs RVQ codebooks
* "RVQ" in general when the `base_v2.py` implementation should be agnostic to either RVQ/FSQ codecs
* the prior mentioned confusion between input/target `quant_level`s
* the mess I dug myself into with AR/NAR/AR+NAR/NAR-len/casual/parallel/"modality" etc.

I should at least have some of it disambiguated whenever I properly document the new implementation with my latest thoughts and observations.

Again, feel free to raise any questions about the project.

**Comment by frank613** [comment]
Thanks for your kind reply again. 

> Oops; in that proposed example, resps_emb.forward(input, name="NAR:1:2", quant_level=2) should be quant_level=1, as the input only has 2 codebooks to embed. Having to mentally do these 0-index / swapping between input/target codebook level confused me there.

I have doubt about this statement because if it is resps_emb.forward(input, name="NAR:1:2", quant_level=1), then we would get the embedding from embeddings[offset+k] = 1 only, which means only the RVQ level 0 code contributes to predict  the RVQ level 2 code, and the code from RVQ level 1 is skipped. I have a feeling that the model being trained with this version of code "skipped" the quant_level = n-1 for modelling quant_level = n. Correct me if I am wrong.

Regards,
Xinwei

**Comment by e-c-k-e-r** [comment]
> I have a feeling that the model being trained with this version of code "skipped" the quant_level = n-1 for modelling quant_level = n. Correct me if I am wrong.

Hm, you might be right. No matter how I try and do the sum loop in my head, it's always off by one.

If it's the case, then I imagine this *might* explain some oddities I've noticed. However, I'm not too sure how much of a detriment it would be to the model, since the model is already somewhat-decent, and its output quality did improve when enabling `cfg.model.experimental.audio_embedding_sums` and some additional post-training. For the model to "learn" to inference later codebooks without needing the immediate prior one wouldn't be all that surprising.

I'll try and validate the code, but my gut says you're right due to the loop being off by one from the discrepancy between `quant_level`. Although, my dev setup might have kicked the bucket overnight, but I'm sure I can use either my personal rig or my 4xV100 training rig to go about it.

**Comment by frank613** [comment]
You're doing awesome, keep going. I Actually am planning to do some non-commercial research based on your code (v1), maybe I should start to look into v2 as well ：）.  BTW, how should I recognise your work if there is something coming out, citation to a paper, for example?

Regards,
Xinwei

**Comment by frank613** [comment]
Oh, I think the direct prior code-level could be important, because according to the essence of RVQ, it is the information residual that the prior level represents being modelled by the next level.  Can't wait to see you release a fixed model!!

> but only when I feel I'm 1000% confident in my understanding of the model and implementation, but that's far down the line.

You are very humble. I learned a lot from your code, it resembles almost every state-of-the-art set-ups of large language models/TTS systems.

Regards,
Xinwei

**Comment by e-c-k-e-r** [comment]
> it is the information residual that the prior level represents being modelled by the next level.

While that is true, it's entirely up to the model itself to decide how it goes both representing the speaker within its latent space, and constructing the final output, but the right path to coerce the model along is unknown (to me at least):
* depriving the model of prior codebook levels removes any dependency it might create when constructing additional codebook levels, and instead reinforces a stronger dependency onto the input prompt itself (rather than possibly flawed prior codebook sequences).
    * As far as I am aware for EnCodec, each residual/codebook is more or less independent from one another, insofar as that the waveform for codebook 0 probably doesn't directly subsequent waveforms from following codebooks.
        * Although, it can be argued that, instead of looking at an RVQ codebook as a fancy Fourier transform, it can be looked at as a sum of tones, and the resultant tone's dependency on its sum does in fact imply there's a relation between its parts, but that's theory that's beyond me (and might not even be necessary in the scope of language modelling where either way is equally learnable).
    * For codebook 0, the model already has to strongly rely on the prompt to construct both its latent representation of the speaker and the final waveform, so there shouldn't be much of a problem for goading the model to do it similarly for prior codebooks.
    * The `ar+nar-len-llama-8` for most of its training only saw the immediately prior codebook for inferencing the next one, so it already demonstrated being able to reconstruct the speaker from little information.
* on the other hand, providing as much detail as possible should always be preferable, even if each subsequent codebook contributes even less detail to the final waveform.
    * Specifically for this, the model does need to try and align each codebook with the previous one. But again, it's enough for the model to be fed just the prior codebook, and everything will align itself without explicitly providing prior codebooks.

But again, I could be entirely wrong as I am still not an expert on any of this, and that's just what I can pull out from past observations at the moment. Like most things with this project, there's multitudes of avenues to go about it, as even most of my "research" seems to amount to it just so happen to working, and replication predicating on a bunch of incidental settings.

I'll still go about doing some post-training fixing whenever I give up on either `nemo-smaller-llama-8` or `nemo-larger-llama-8` (whichever ~~disappoints me more~~ proves to be less fruitful) to free up a rig, but since the model didn't seem to have any fundamental differences after correcting it, I do expect it to be negligible as the model already was coerced into the way it is.

> You are very humble. I learned a lot from your code, it resembles almost every state-of-the-art set-ups of large language models/TTS systems.

Glad there's something to be gleaned from my madness.

Honestly, I find that a lot of papers and implementations for other solutions leave out the "why" behind a lot of things that might seem insignificant. Not to say I'm not guilty of throwing things in without understanding the rhyme or reason behind it, but I feel I at least eventually figure out the "why" and note it. If I do remember right, TorToiSe's "paper" / doc page tries to go about its entire stack to some degree, but it's been a long while since I've last looked at it.

**Comment by e-c-k-e-r** [comment]
> Is it because it puts more weights on level 0 and level 1 code for sampling a training batch?

For the old implementation: implicitly. It's "weighed" by virtue of assignined the target codebook to train each batch under, and lower codebooks are prioritized way more, thus having the model focus more on them over higher codebooks simply from being exposed to it more. I did find that having each codebook equally trained seemed to harm the model, but I wouldn't be surprised if some other factor was causing that (and the new implementation doesn't seem to have this problem).

> Can you please share the origin of the theory.

It's just something that keeps coming to mind when I mull over both the model's architecture and my mental model of how EnCodec works. I wouldn't really have a good idea on how to go about testing it without having to do deep dives on actual papers on speech or music or the physics of sound and what-not. At the very least, testing if there's a substantial dependency between each codebook (/ fundamental wave in the final sound).

> where the author has found that using the code-books' representations as initialisation of the embeddings could help a lot. I think it makes the model more aware of the acoustic than pure language modelling of the codes.

If you mean by (from the linked paper):

> We can either initialize the weights of the embedding model randomly or use the
learnt codebooks from the trained NAC model to provide a better starting point

that they just used the embeddings from EnCodec itself, I did have code for that in the past to source external embeddings:

https://github.com/e-c-k-e-r/vall-e/blob/be83ddabaa32870ff3ce1092d000fb12a13184aa/vall_e/models/base.py#L233)

but it was quite clunky, and if I recall, it didn't prove much fruitful.

I could probably experiment with pre-initializing the embeddings for the new implementation if I train from scratch again, as the initial audio embeddings' width does not need to be tied to the underlying model's width itself. But I feel it's might only marginally speed up the initial phase of training for speech to emerge.

I'll read through the paper if I get any freetime soon. Despite the `stt` task being neglected for the old implementation (it really did not perform to my expectation, but it was a proof of concept feature I glued on) and the new one not supporting it yet, it's still a feature that interests me, and I can see if there's anything to glean from it.

**Comment by e-c-k-e-r** [comment]
> Can't wait to see you release a fixed model!!

I've freed up my 7900XTX rig so I could perform more post-training for the `ar+nar-len-llama-8` reference model. The weights should be up already, although I still need to do some evals to see if there's any fundamental differences to the outputs with it properly using all prior RVQ levels. Additionally, it should have been trained against:
* noise suppression (`ns`)
* speech removal (`sr`)
* "No phonemes" mode (uses `raw_text_embs` instead of `text_embs`, since the new implementation ironically fixed the old implementation with this mode)

Some caveats:
* post-training was for half the epoch for my utterances between two seconds and twelve, for speakers with at least eight utterances
    * there *may* be a regression for outputs above twelve seconds
* half of this post-training session was purely training the `NAR-len` modality, then with a 90%/10% `NAR-len`/`AR+NAR`
    * Although I'm treating `AR+NAR` modality as a second class feature now, it *somewhat* still works.
* partially with Prodigyopt, then AdamW-8bit with a learning rate of `5.0e-5` (as my 7900XTX rig had a new instability emerge with some NCCL/CUDA calloc failing on checkpoint saving some of the time)

**Comment by frank613** [comment]
Thanks a lot for the update!

Regards,
Xinwei

**Comment by e-c-k-e-r** [comment]
Incidentally following up on the "pre-init the audio embeddings with the ones from the NAC" idea (partially for posterity, as I don't think I have a home for these notes specifically): it doesn't seem all that fruitful.

While I am 50% confident I probably botched it somewhere (since I'm reusing the old code I have for retrieving DAC's embeddings as it makes things easier, and there were regressions I had to work around), none of my experiments the past week shows a tangible benefit over no explicit initialization at best and divergence at worse. I will caveat that half of my tests were called off around 20k steps (batch size of 32) if I feel the loss curve isn't adequate, as I don't think it's worth dolling out a day to something that progresses similarly to the first failed experiment.
* I will also note that the bandaid fixes for the regressions I had (from `all_groups_norm > 0` asserts failing from gradient issues, to convergence problems) seemed to have stemmed from `AudioEncoder.level_weights` being a detriment, despite prior experiments for Encodec/DAC/`nvidia/audio-codec-44khz` not showing these problems, even under bfloat16 on my 7900XTX rig. Honestly I feel it's not necessary to have it, as either the embeddings themselves or the encoder's feed forward would handle this anyways.

I might need to re-read the paper again in the event I missed a detail, but I imagine this would only *really* work for simpler audio-oriented tasks such as ASR, rather than synthesis (something => audio or audio => something, rather than something + audio => audio). I feel this might be more targeted against VALL-E (or my implementation, or any other solution that doesn't rely on exotic feature extractions) as the embeddings/features for a NAC are fundamentally different to the ones that VALL-E (and the like) would require to function. My prior failings suggest that it's really easy to have problems crop up from doing something funny with the embeddings.

**Comment by frank613** [comment]
Thanks a lot for the update. I also had a feeling that paper was only targeting ASR. 

Best,
Xinwei
